simulator/src/models/power-plant.ts

```diff
@@ -0,0 +1,450 @@
+        if (rows.length == 1) {
+            let row = rows[0];
+            let plant = new PowerPlant(row.id,
+                            row.owner,
+                            row.start_delay,
+                            row.stop_delay,
+                            row.production_level,
+                            row.production_capacity,
+                            row.production_variant,
+                            row.production_ratio,
+                            row.battery_capacity,
+                            row.created_at,
+                            row.updated_at);
+            return plant
+        } else {
+            return Promise.reject("Could not find any wind object with id " + id);
+        }
+    }
+
+    /**
+     * Store a given power plant data in the power_plant_data table.
+     * If there already exists data for this time then update instead.
+     */
+    store() {
+        ElectricityGridDB.table('power_plant').insert_or_update({
+            id: this._id,
+            owner: this._owner,
+
+            start_delay: this.startDelay,
+            stop_delay: this.stopDelay,
+
+            production_level: this._productionLevel,
+            production_capacity: this.productionCapacity,
+            production_variant: this.productionVariant,
+            production_ratio: this._productionRatio,
+        
+            battery_capacity: this._battery.capacity,
+
+            created_at: this.createdAt,
+            updated_at: this.updatedAt,
+        }, ['id']);
+    }
+
+
+    /**
+     * Starts the power plant.
+     * @note Takes some time before its starts.
+     * @note The power plant can only start if its Status is equal Stopped.
+     */
+    start() {
+        if (this.status == Status.Stopped) {
+            this.status = Status.Starting;
+            setTimeout(() => {
+                this.status = Status.Running;
+            }, this.startDelay);
+        }
+    }
+
+
+    /**
+     * Stops the power plant.
+     * @note Takes some time before its stops.
+     * @note The power plant can only stop if its Status is equal to Running.
+     */
+    stop() {
+        if (this.status == Status.Running) {
+            this.status = Status.Stopping;
+            setTimeout(() => {
+                this.status = Status.Stopped;
+            }, this.stopDelay);
+        }
+    }
+
+
+    /**
+     * Update the power plant data in the database.
+     * This method should be called each simulation checkpoint to ensure that
+     * the database is populated with new data.
+     * @param {Simulation} sim the simulation instance.
+     */
+    update(sim: Simulation) {
+        let time = sim.time;
+
+        (async () => {
+            await storePowerPlantData({id: this._id,
+                time: time,
+                production: this.deltaProduction,
+                battery_value: this._battery.value,
+                unit: this.unit});
+        })();
```

Comment:
Do we need to store historical data every update? This runs every second resulting in lots of data storing.
I thought the update(sim) only carries out the calculations for determining the current values, I may have confused you about the update method, but step calls the update method via the state class.

Reply:
Do we need historical data? That was only an advanced requirement for wind speeds.

Reply:
We don't need it but it counts as an advanced requirement and should not be too difficult to implement.